src/lib/index/wb_cache.cpp

```diff
@@ -431,32 +485,103 @@ void IndexWBCache::recover(sisl::byte_view sb) {
                         m_vdev->commit_blk(buf->m_blkid);
                     }
                     l0_bufs.push_back(buf);
+                    // insert upbuffer and its upbuffer to the recovering_upbuffers set until nullptr
+                    // suppose the following scenario that crash happened when flushing B
+                    //    ├── A (WAITING FOR B)
+                    //    │    ├── B (CRASHED)
+                    //    │    │    ├── C (FLUSHED)
+                    //    │    │    └── D (FLUSHED)
+                    //    │    │         └── E (FLUSHED)
+                    //    │    └── F (FLUSHED)
+                    //    └── G (FLUSHED)
+                    // we need to recover B and A. During recover_buf(A), we need to recover F and commit its blk. So in
+                    // this case, we give a second chance to new buffers to be a part of repair. There is still a case
+                    // that committing new buffers is not needed since they are not part of recovery paths.
+                    auto up_buf = buf->m_up_buffer;
+                    while (up_buf) {
+                        recovering_upbuffers.insert(up_buf);
+                        up_buf = up_buf->m_up_buffer;
+                    }
                 } else {
-                    buf->m_up_buffer->m_wait_for_down_buffers.decrement();
+                    // there is a chance it is a second tier buffer, so we need to repair it in the second pass
+                    if (!was_node_committed(buf->m_up_buffer)) {
+                        second_tier_buffers.insert(buf);
+                    } else {
+                        buf->m_up_buffer->m_wait_for_down_buffers.decrement();
+#ifndef NDEBUG
+                        bool found{false};
+                        for (auto it = buf->m_up_buffer->m_down_buffers.begin();
+                             it != buf->m_up_buffer->m_down_buffers.end(); ++it) {
+                            auto sp = it->lock();
+                            if (sp && sp == buf) {
+                                found = true;
+                                buf->m_up_buffer->m_down_buffers.erase(it);
+                                break;
+                            }
+                        }
+                        HS_DBG_ASSERT(found,
+                                      "Down buffer is linked to Up buf, but up_buf doesn't have down_buf in its list");
+#endif
+                    }
+                }
+            }
+        }
+    }
+    // remove the second tier buffers from the recovering_upbuffers if they are in it and upgrade their up buffers
+    for (auto const& buf : second_tier_buffers) {
+        // if buf->m_up_buffers is in recovering_upbuffers, commit the buf  and add it to l0_bufs o.w, remove the
+        // down_count of upbuffer
+        if (recovering_upbuffers.find(buf->m_up_buffer) != recovering_upbuffers.end()) {
+            m_vdev->commit_blk(buf->m_blkid);
+            l0_bufs.push_back(buf);
```

Comment:
What if `buf` itself is in the `recovering_upbuffers`? This might cause duplicate decrements to the `down_buf` counter

Reply:
buf is new node, so it cannot be included in up buffers. Did I miss something?